### ISSUE TITLE
Cache does not seem to update correctly if we rename `.clj` files to `.cljc`

### ISSUE THREAD

**Issue by andrestylianos** [clj-kondo/clj-kondo#1964]
**version**
v2023.01.16

**platform**
native

**problem**

From what I could understand of my issue, it seems like `clj-kondo` does not correctly invalidate the cache when we rename `.clj` files to `.cljc`.

**repro**

This involves the cache and more than one file, so the minimal repro I came up with is:

We have a file,`src/repro/main.clj`, that requires another file, `src/repro/utils.clj`. 

```clojure
(ns repro.utils)

(defn example [a b]
  (+ a b))
```

```clojure
(ns repro.main
  (:require [repro.utils :as utils]))

(utils/example 1 2)
```

lint this with `clj-kondo --lint src` (using a cache) gives us
```linting took 24ms, errors: 0, warnings: 0```

If we then rename `src/repro/utils.clj` to `.cljc` and make some changes, kondo won't pick up those changes unless we wipe the cache:

```clojure
(ns repro.utils)

(defn example [a b c]
  (+ a b c))
```

```clojure
(ns repro.main
  (:require [repro.utils :as utils]))

(utils/example 1 2 3)
```

After this, linting again gives us:
```
src/repro/main.clj:4:1: error: repro.utils/example is called with 3 args but expects 2
linting took 20ms, errors: 1, warnings: 0
```

**expected behavior**

I expected `clj-kondo` to give no errors the second time as well, but instead it seems to be using cached information from the `.clj` file that does not exist anymore.

**Comment by borkdude** [comment]
I'm not sure if this can be support reliably since I don't know how clj-kondo could detect if a namespace was removed or renamed. Clj-kondo doesn't really scan the filesystem for relations between files: it just lints a file and saves the namespace state or overwrites an existing namespace. Since you can have both .clj and .cljc for the same namespace, it saves both under different entries.

For now it might be better to just clear the cache with `rm -r .clj-kondo/.cache` in this case, however, I understand that that's not the best UX. I'll leave this issue open to think more about it. Perhaps clojure-lsp can also play a role here (cc @ericdallo ).

**Comment by ericdallo** [comment]
yeah, that sounds tricky to handle, if we go with the remove cache approach, when we would remove? during startup? during every lint? sounds a little bit weird to me

**Comment by colin-p-hill** [comment]
Just came here to report the same issue, with the same manifestation as an arity error. Invalidating the cache would be hard, but I'm not sure if that's the central problem; this issue can in theory arise without a stale cache.

```clojure
;; repro/repro.clj
(ns repro)
(defn my-fn [_ _] nil)
(my-fn :foo :bar)
```
```clojure
;; repro/repro.cljs
(ns repro)
(defn my-fn [_] nil)
(my-fn :foo)
```
```clojure
;; repro/repro.cljc
(ns repro)
(defn my-fn [] nil)
(my-fn)
```
```text
> clj-kondo --cache false --lint repro
repro\repro.cljc:3:1: error: repro/my-fn is called with 0 args but expects 2
repro\repro.cljc:3:1: error: repro/my-fn is called with 0 args but expects 1
linting took 14ms, errors: 2, warnings: 3
```
It seems like clj-kondo is checking the call against all dialect-specific definitions and disregarding the cljc one, even in the same file. Maybe the thing that needs a tweak here is the way clj-kondo handles a var that has multiple definitions.

**Comment by borkdude** [comment]
@colin-p-hill That seems to be a different issue, having to do with how clj-kondo resolves a var from `.cljc` + `.clj` namespaces. My guess is that the same error would happen if you leave out the `.cljs` file, but I haven't checked. Can you post a new issue about this?
